tests/test_tracer.py

```diff
@@ -2,551 +2,424 @@
-
-    def __init__(self):
-        # original call
-        super(DummyWriter, self).__init__()
-        # dummy components
-        self.spans = []
-        self.traces = []
-        self.services = {}
-        self.json_encoder = JSONEncoder()
-        self.msgpack_encoder = MsgpackEncoder()
-
-    def write(self, spans=None, services=None):
-        if spans:
-            # the traces encoding expect a list of traces so we
-            # put spans in a list like we do in the real execution path
-            # with both encoders
-            trace = [spans]
-            self.json_encoder.encode_traces(trace)
-            self.msgpack_encoder.encode_traces(trace)
-            self.spans += spans
-            self.traces += trace
-
-        if services:
-            self.json_encoder.encode_services(services)
-            self.msgpack_encoder.encode_services(services)
-            self.services.update(services)
-
-    def pop(self):
-        # dummy method
-        s = self.spans
-        self.spans = []
-        return s
-
-    def pop_traces(self):
-        # dummy method
-        traces = self.traces
-        self.traces = []
-        return traces
-
-    def pop_services(self):
-        # dummy method
-        s = self.services
-        self.services = {}
-        return s
 
+class TracerTestCase(BaseTracerTestCase):
+    def test_tracer_vars(self):
+        span = self.trace('a', service='s', resource='r', span_type='t')
+        span.assert_matches(name='a', service='s', resource='r', span_type='t')
+        # DEV: Finish to ensure we don't leak `service` between spans
+        span.finish()
 
-def get_dummy_tracer():
-    tracer = Tracer()
-    tracer.writer = DummyWriter()
-    return tracer
-
-
-def test_default_hostname_from_env():
-    # it should use default hostname from DATADOG_TRACE_AGENT_HOSTNAME if available
-    try:
-        with mock.patch.dict('os.environ', {'DATADOG_TRACE_AGENT_HOSTNAME': 'customhost'}):
-            del sys.modules['ddtrace.tracer']  # force reload of module
-            from ddtrace.tracer import Tracer
-            eq_('customhost', Tracer.DEFAULT_HOSTNAME)
-    finally:
-        del sys.modules['ddtrace.tracer']  # clean up our test module
-        from ddtrace.tracer import Tracer
-        eq_('localhost', Tracer.DEFAULT_HOSTNAME)
+        span = self.trace('a')
+        span.assert_matches(name='a', service=None, resource='a', span_type=None)
+
+    def test_tracer(self):
+        def _mix():
+            with self.trace('cake.mix'):
+                pass
+
+        def _bake():
+            with self.trace('cake.bake'):
+                pass
+
+        def _make_cake():
+            with self.trace('cake.make') as span:
+                span.service = 'baker'
+                span.resource = 'cake'
+                _mix()
+                _bake()
+
+        # let's run it and make sure all is well.
+        self.assert_has_no_spans()
+        _make_cake()
+
+        root = self.get_root_span()
```

Comment:
nothing happens with `root`?

Reply:
whoops, this was probably back before I had `self.assert_structure()` it used to be:

```python
root = self.get_root_span()
root.assert_structure( ... )
```

Reply:
oh, nvm, `root` is being used, but I updated to be more clear